Mathlib/Data/Nat/Digits/Lemmas.lean

```diff
@@ -137,6 +139,116 @@ theorem base_pow_length_digits_le (b m : ℕ) (hb : 1 < b) :
+    2 * ∑ n ∈ Finset.range (b^k), (b.digits n).sum = (b-1) * k * b^k := by
+  by_cases k0 : k = 0
+  · simp [mul_comm, k0]
+  by_cases hb : b = 0 ∨ b = 1
+  · rcases hb with (h | h) <;> simp [h, k0]
+  replace hb : 1 < b := by omega
+  have sum_sum_digits : 2 * ∑ x ∈ Finset.range b, (b.digits x).sum = b * (b - 1) := by
+    rw [mul_comm, <- Finset.sum_range_id_mul_two]
+    congr! with x x_in
+    simp at x_in
+    simp [digits_sum_eq_self x_in]
+  induction' k, (by omega : 1 ≤ k) using Nat.le_induction with m pos_m ih
+  · simp [sum_sum_digits, mul_comm]
+  · specialize ih (Nat.ne_zero_of_lt pos_m)
+    have digits_eq_cons := Nat.digits_eq_cons_digits_div
+      hb (by positivity)
+    have pos_n : ∀ n ∈ Finset.Ico 1 (b^m), 0 < n := by
+      intro n n_in
+      simp at n_in
+      exact n_in.left
+    have sum_pow_add_eq_sum_pair {f : ℕ → ℕ} {i j b : ℕ} :
+      ∑ x ∈ Finset.range (b^(i+j)), f x =
+      ∑ p ∈ Finset.range (b^i) ×ˢ Finset.range (b^j),
+        f (b^j * p.fst + p.snd) := by
+      have sum_range_mul_eq_sum_pair (i j) : ∑ x ∈ Finset.range (i*j), f x =
+        ∑ p ∈ Finset.range i ×ˢ Finset.range j, f (j * p.fst + p.snd) := by
+        have sum_add_eq_sum_pair : ∑ x : Fin (i * j), f x =
+          ∑ p : Fin i × Fin j, f (j * p.fst + p.snd : ℕ) := by
+          simp [add_comm, <- Equiv.sum_comp (e := finProdFinEquiv)]
+        simp [sum_add_eq_sum_pair, Finset.sum_range, Fintype.sum_prod_type,
+          Finset.sum_product]
+      convert sum_range_mul_eq_sum_pair (b^i) (b^j) using 1
+      rw [<- Nat.pow_add]
+    rw [sum_pow_add_eq_sum_pair]
+    have : ∑ p ∈ Finset.range (b ^ m) ×ˢ Finset.range b, (Nat.digits b (b * p.fst + p.snd)).sum =
+        ∑ p ∈ Finset.range (b ^ m) ×ˢ Finset.range b,
+        ((Nat.digits b (b * p.fst)).sum + (Nat.digits b p.snd).sum) := by
+      apply Finset.sum_congr rfl
+      intro p hp
+      have h : p.snd < b := by
+        rw [Finset.mem_product] at hp
+        exact Finset.mem_range.mp hp.right
+      have digit_sum_add_single_digit (n d : ℕ) (hd : d < b) :
+        (Nat.digits b (b * n + d)).sum = (Nat.digits b (b * n)).sum + (Nat.digits b d).sum := by
+        rcases d
+        · simp
+        · rw [add_comm, Nat.digits_add b hb _ _ hd (by omega)]
+          simp [add_comm, List.sum_cons, digits_sum_eq_self hd]
+          symm
+          exact digits_sum_base_mul_cancel (Nat.zero_lt_of_lt hb)
+      exact digit_sum_add_single_digit p.1 p.2 h
+    simp [this]
+    clear this
+    rw [@Finset.sum_product]
+    simp [Finset.sum_add_distrib, mul_add, <- Finset.mul_sum]
+    conv_lhs => rhs; rw [<- mul_assoc, mul_comm 2, mul_assoc, sum_sum_digits]
+    simp [digits_sum_base_mul_cancel (Nat.zero_lt_of_lt hb)]
+    rw [<- mul_assoc, mul_comm 2, mul_assoc, ih]
+    ring
+
+/-- A specialization of `sum_digit_sum_base_pow_eq` to base 10.
+-/
+theorem sum_digit_sum_ten_pow_eq {k : ℕ} :
+  (∑ n ∈ Finset.range (10^k), (Nat.digits 10 n).sum) =
+    45 * k * 10^(k-1) := by
```

Comment:
you could probably have an intermediate lemma in this form, i.e. where the RHS would be `((b - 1) * b) / 2 * k * b ^ (k - 1)`.